### ISSUE TITLE
Server Graceful Shutdown

### ISSUE THREAD

**Issue by tnewman** [Eugeny/russh#104]
Currently, russh doesn't provide a graceful shutdown mechanism. Graceful shutdowns are especially important in cloud environments to allow instances to be shut down without impacting users.

The following should happen on shutdown:
1. Stop listening accepting new connections
2. Notify application code that a shutdown has been triggered
3. Allow application code to notify that shutdown has been complete
4. Wait for the existing connections to close
5. If application code misbehaves and doesn't close the connection within a certain period of time, close the connection

**Comment by Eugeny** [comment]
This is not directly in the scope of this library.

> Stop listening accepting new connections

Stopping polling and dropping the future returned by `russh::server::run` does this. Example (quickly cobbled together, pseudocode):

```rust
let (rx, tx) = mpsc::channel(1);

tokio::select! {
   rx.recv() => (),
   russh::server::run(config, &std::net::SocketAddr::from_str("0.0.0.0:2222").unwrap(), server)
}
```

> Notify application code that a shutdown has been triggered
> Allow application code to notify that shutdown has been complete

You can use the same method of passing an abort signal channel to your `Handler` structs.

> Wait for the existing connections to close

Currently, outstanding sessions are not waited on when the server is stopped, but you can implement your own `run` using `run_stream` and keep reference to the `JoinHandle`s for each session so that you can wait for them later: https://docs.rs/russh/latest/src/russh/server/mod.rs.html#610-628

> If application code misbehaves and doesn't close the connection within a certain period of time, close the connection

Same but add an abort signal channel to `run_stream` like in #1.

**Comment by RLuzi** [comment]
Hi, sorry to bother you. I'm trying to implement a graceful shutdown mechanism with russh version 0.51.1 and I don't understand how to take a JoinHandle reference, as suggested in the previous comment. Could you please confirm that this is still possible? The link suggested in the previous comment seems to lead to some code that is not related to the topic at hand.

**Comment by Eugeny** [comment]
Yes - to still can do it by reimplementing the `Server::run_on_socket` method yourself, which allows you to work with the `RunningSession` / `JoinHandle` structs directly.

**Comment by Eugeny** [comment]
I meant just copy pasting it and adjusting for the behaviour you want to happen at the shutdown

**Comment by Eugeny** [comment]
Sounds like #495 which was fixed in v0.52.0

**Comment by RLuzi** [comment]
hello, is there any news on this topic, also considering my last comment regarding the PR you linked?

**Comment by Eugeny** [comment]
Sorry - I've missed your comment, I'll try to look at it in the coming days

**Comment by RLuzi** [comment]
Hi, is there any news on this topic? Can I help you in any way?

**Comment by Eugeny** [comment]
I've added built-in graceful shutdown in #539, but if you'd like to do it on the stable release, you can just copy-paste the implementation of `run_on_socket` from there
